grpc-circuitbreaker-utils/src/main/java/org/hypertrace/circuitbreaker/grpcutils/resilience/ResilienceCircuitBreakerInterceptor.java

```diff
@@ -0,0 +1,124 @@
+package org.hypertrace.circuitbreaker.grpcutils.resilience;
+
+import io.github.resilience4j.circuitbreaker.CircuitBreaker;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import org.hypertrace.circuitbreaker.grpcutils.CircuitBreakerConfiguration;
+import org.hypertrace.circuitbreaker.grpcutils.CircuitBreakerInterceptor;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+@Slf4j
+public class ResilienceCircuitBreakerInterceptor extends CircuitBreakerInterceptor {
+
+  private final ResilienceCircuitBreakerProvider resilienceCircuitBreakerProvider;
+  private final CircuitBreakerConfiguration<?> circuitBreakerConfiguration;
+  private final Clock clock;
+
+  ResilienceCircuitBreakerInterceptor(
+      CircuitBreakerConfiguration<?> circuitBreakerConfiguration,
+      Clock clock,
+      ResilienceCircuitBreakerProvider resilienceCircuitBreakerProvider) {
+    this.circuitBreakerConfiguration = circuitBreakerConfiguration;
+    this.clock = clock;
+    this.resilienceCircuitBreakerProvider = resilienceCircuitBreakerProvider;
+  }
+
+  @Override
+  protected boolean isCircuitBreakerEnabled() {
+    return circuitBreakerConfiguration.isEnabled();
+  }
+
+  @Override
+  protected <ReqT, RespT> ClientCall<ReqT, RespT> createInterceptedCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    return new ForwardingClientCall.SimpleForwardingClientCall<>(
+        next.newCall(method, callOptions)) {
+      CircuitBreaker circuitBreaker;
+      String circuitBreakerKey;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        Instant startTime = clock.instant();
+        // Wrap response listener to track failures
+        Listener<RespT> wrappedListener =
+            wrapListenerWithCircuitBreaker(responseListener, startTime);
+        super.start(wrappedListener, headers);
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public void sendMessage(ReqT message) {
+        CircuitBreakerConfiguration<ReqT> config =
+            (CircuitBreakerConfiguration<ReqT>) circuitBreakerConfiguration;
+        // Type check for message class compatibility
+        if (config.getRequestClass() != null && !config.getRequestClass().isInstance(message)) {
```

Comment:
This doesn't need to hold. Treat the request class as a filter. If I add an interceptor to a stub for one method of the stub, the other methods should not be circuit broken. Since we can't add the calls more granularly than the stub level, we'll never have interceptors that only are being called for one specific class.

Reply:
Couldn't understand this, what need to be done here ?

Reply:
Responded offline, but basically this request class is a criteria to be met before we assign a circuit breaker. If it's not met, we should skip the circuit breaker rather than erroring. This is because we attach interceptors at the stub or channel level, so if the circuit breaker is for a single method we want it to ignore all other methods. If it's for multiple methods, it shouldn't be keyed by the request (although we could support multiple methods keying off an aspect of the context like tenant).

Reply:
Fixed it